### ISSUE TITLE
no-param-reassign should trigger loops parameters

### ISSUE THREAD

**Issue by antoine-pous** [eslint/eslint#13909]
**What rule do you want to change?**
```json
{
  "no-param-reassign": ["error", {
    "props": false
  }]
}
```

**Does this change cause the rule to produce more or fewer warnings?**
More, but it must be opt-out by adding a new rule parameter
```json
{
  "no-param-reassign": ["error", {
    "props": false
    , "loops": true
  }]
}
```

**Please provide some example code that this change will affect:**
```ts
const arrObj = [{name: 1},{name: 2},{name: 3}]
for(const obj in arrObj) {
  obj.name = "anything else" // With the new rule option this shouldn't be possible
}
```

**What does the rule currently do for this code?**
Nothing, it seem to be ignored

**What will the rule do after it's changed?**
Trigger an error or a warn if the rule is opt-in with `loops` parameter

**Are you willing to submit a pull request to implement this change?**
Nope, i unfortunately don't have the time :/

**Comment by mdjermanovic** [comment]
Hi @antoine-pous, thanks for the issue!

`no-param-reassign` targets function parameters. I don't think this rule should be checking anything that isn't a parameter, like loop variables. 

> ```ts
> const arrObj = [{name: 1},{name: 2},{name: 3}]
> for(const obj in arrObj) {
>   obj.name = "anything else" // With the new rule option this shouldn't be possible
> }
> ```

I'm curious about this example, was it a typo (for-in loop instead of for-of loop), or a proposal to catch assignments to properties of a `string` value?

**Comment by antoine-pous** [comment]
Yeah i would like catch the properties assignment too. IMO it must be considered as a mutation, and should be part of the `no-param-reassign` due to the possibility to perform that:

```json
{
    "no-param-reassign": ["error", {
        "props": true
    }]
}
```

```ts
const exec = (params: {name: number | string}[]): void => {
  for(const param of params) {
    param.name = 'muted'
  }
}

exec([{name: 1} ,{name: 2} ,{name: 3}])
```

**Comment by anikethsaha** [comment]
if we take the js example for the above, and if we want to support it, I think this would require changes and support from `eslint-scope` as it doesn't support props as of now.

**Comment by mdjermanovic** [comment]
I'm not sure if it's common to disallow mutations of for-of loop variables. There are some examples in ESLint's code:

https://github.com/eslint/eslint/blob/06b58096975935ec016d96dd5f333f059c270f26/lib/eslint/eslint.js#L390-L392

As for the parameters, `no-param-reassign` with `"props": true` can catch only modifications made directly through the param reference. We can't really track all local variables, assignments, and references across the function to make sure that the object will be read-only. Also, in `for (foo of bar)`, `foo` doesn't have to be a property of `bar` if `bar` isn't an array.

**Comment by apieceofbart** [comment]
Hey, I basically have a similar case - I was bitten today by mutating props in react component. Simplified code:
```
function MyFunc(props) {
const attribute = props.attributes.find(...NOT IMPORTANT)
attribute.foo = 'bar'
}
```

This looks like a simple case but was already too complex for the rule to catch. I think if it's not possible to make the rule catch those assignments then it would be good to add some short warning in the docs.
